README.md
## Pnrhost Domain Registrar SDK

CHANGELOG.md
## <small>3.0.2 (2021-01-29)</small>

* Removed node_modules from git ([578b76b](https://github.com/privyreza/pdr-php-sdk/commit/578b76b))
* fix: minor serach fix ([7fc4e94](https://github.com/privyreza/pdr-php-sdk/commit/7fc4e94))



## <small>3.0.1 (2021-01-29)</small>

* Reverted domain searches ([3f278c0](https://github.com/privyreza/pdr-php-sdk/commit/3f278c0))



## 3.0.0 (2021-01-28)

* Fixed the search domain function ([07fa5b2](https://github.com/privyreza/pdr-php-sdk/commit/07fa5b2))
* Upgraded guzzle ([3014f10](https://github.com/privyreza/pdr-php-sdk/commit/3014f10))
* fix: changed domain-serarches to searches ([5434beb](https://github.com/privyreza/pdr-php-sdk/commit/5434beb))



## <small>2.0.1 (2020-12-27)</small>

* Added ability to cc consultant ([be7b0fb](https://github.com/privyreza/pdr-php-sdk/commit/be7b0fb))
* Added fallback for host ([dc228ff](https://github.com/privyreza/pdr-php-sdk/commit/dc228ff))



## 2.0.0 (2020-11-17)

* Added ability to copy consultant on all emails from zispa ([2a0816a](https://github.com/privyreza/pdr-php-sdk/commit/2a0816a))



## <small>1.1.1 (2020-09-01)</small>

* Release 1.1.1 ([a6cb992](https://github.com/privyreza/pdr-php-sdk/commit/a6cb992))
* Release 1.1.1 ([70472ce](https://github.com/privyreza/pdr-php-sdk/commit/70472ce))
* Release 1.1.1 ([818fcff](https://github.com/privyreza/pdr-php-sdk/commit/818fcff))
* fix: Removed extra data referrence on registerDomain and transferDomain function ([868f980](https://github.com/privyreza/pdr-php-sdk/commit/868f980))



## 1.1.0 (2020-08-22)

* feat: Added getDomains function ([50968a5](https://github.com/privyreza/pdr-php-sdk/commit/50968a5))
* feat: Added the transferDomain function ([f697381](https://github.com/privyreza/pdr-php-sdk/commit/f697381))



## <small>1.0.3 (2020-06-26)</small>

* fix: Resolved issue when checking remote domains ([c51103f](https://github.com/privyreza/pdr-php-sdk/commit/c51103f))



## <small>1.0.2 (2020-06-26)</small>

* Upgraded vlucas/phpdotenv to 4.1.7 ([abee378](https://github.com/privyreza/pdr-php-sdk/commit/abee378))



## <small>1.0.1 (2020-06-24)</small>

* Release 1.0.1 ([7b796b0](https://github.com/privyreza/pdr-php-sdk/commit/7b796b0))
* Release 1.0.1 ([1856ff7](https://github.com/privyreza/pdr-php-sdk/commit/1856ff7))
* Release 1.0.1 ([3c3bea8](https://github.com/privyreza/pdr-php-sdk/commit/3c3bea8))
* fix: corrected the CHANGLOG.md ([9b48e1b](https://github.com/privyreza/pdr-php-sdk/commit/9b48e1b))
* fix: Fixed error while getting errors proprty from guzzle 201 or 200 response ([1688432](https://github.com/privyreza/pdr-php-sdk/commit/1688432))
* fix: Removed other contacts for now - tech ([a2e74d9](https://github.com/privyreza/pdr-php-sdk/commit/a2e74d9))



## <small>1.0.1 (2020-06-24)</small>

* Release 1.0.1 ([1856ff7](https://github.com/privyreza/pdr-php-sdk/commit/1856ff7))
* Release 1.0.1 ([3c3bea8](https://github.com/privyreza/pdr-php-sdk/commit/3c3bea8))
* fix: corrected the CHANGLOG.md ([9b48e1b](https://github.com/privyreza/pdr-php-sdk/commit/9b48e1b))
* fix: Fixed error while getting errors proprty from guzzle 201 or 200 response ([1688432](https://github.com/privyreza/pdr-php-sdk/commit/1688432))
* fix: Removed other contacts for now - tech ([a2e74d9](https://github.com/privyreza/pdr-php-sdk/commit/a2e74d9))



## 1.0.0 (2020-04-25)

* Added functionality for NS update ([8c37fea](https://github.com/privyreza/pdr-php-sdk/commit/8c37fea))
* Added Licence file ([ad3db9b](https://github.com/privyreza/pdr-php-sdk/commit/ad3db9b))
* Added minimal composer.json in preparation for packagist publish ([9a0c653](https://github.com/privyreza/pdr-php-sdk/commit/9a0c653))
* Added more exception handling ([3d9fb5c](https://github.com/privyreza/pdr-php-sdk/commit/3d9fb5c))
* Changed conflicting vlucas/phpdotenv version ([46a0fca](https://github.com/privyreza/pdr-php-sdk/commit/46a0fca))
* created all registration function and a test index file ([57fb672](https://github.com/privyreza/pdr-php-sdk/commit/57fb672))
* Initial Commit ([9c2126b](https://github.com/privyreza/pdr-php-sdk/commit/9c2126b))





index.php
<?php
require "vendor/autoload.php";

$dotenv = Dotenv\Dotenv::createImmutable(__DIR__);
$dotenv->load();

use Pdr\Client;

$client = new Client();

// print_r($client);

// Search domain
$domain = 'ipokkkll-wer.co.zw';
$domainsearch = $client->searchDomain($domain);

$status = $domainsearch->data->attributes->status;

if($status === 'available') {
    // Build post data
    $postfields = array(
        'domain' => $domain,
        'nameservers' => array(
            "ns1" => "ns1.pnrhost.com",
            "ns2" => "ns2.pnrhost.com"
        ),
        'contacts' => array(
            'registrant' => array(
                "first_name" => "SDK Privy",
                "last_name" =>"SDK Reza",
                "email" =>"privyreza@gmail.com",
                "company" =>"Pnrhost",
                "mobile" =>"0773234827",
                "street_address" =>"78 Test Street",
                "core_business" =>"SDK Test Core Business",
                "city" =>"Harare",
                "country" =>"Zimbabwe"
            ),
        )
    );
    call('RegisterDomain', $postfields);
}

function call($action, $data){
    global $client;
    $func =  lcfirst($action);

    return $client->$func($data);
}

src/Client.php
<?php
namespace Pdr;

use GuzzleHttp\Psr7\Request;
use GuzzleHttp\Exception\RequestException;
use GuzzleHttp\Exception\ClientException;
use GuzzleHttp\Exception\ServerException;

class Client
{
    /**
     * The client for making http calls to api
     */
    protected $client;

    /**
     *  The API host URL
     * 
     */
    protected $host;

    /**
     * Api Url 
     */
    protected $apiUrl;

    /**
     * The API version to connect to
     */
    protected $apiVersion;

    /**
     * token
     */
    protected $token;

    /**
     * Request headers
     */
    protected $headers;

    /**
     * Set env
     */
    protected $debug;

    /**
     * Initialize the client
     */
    public function __construct($token){
        $this->host = getenv('HOST');
        // If host is null use api.resellme.co.zw
        if (is_null($this->host)) {
            $this->host = 'https://api.resellme.co.zw';
        }
        
        $this->apiVersion = getenv('API_VERSION');
        $this->apiUrl = $this->host . '/api/' . $this->apiVersion;
        $this->debug = getenv('DEBUG');

        $this->client = new \GuzzleHttp\Client(
            [
                "base_uri" => $this->host
            ]
        );

        // Set Headers
        $this->headers = [
            "Accept" => "application/vnd.api+json",
            "Content-Type" => "application/vnd.api+json",
            'Authorization' => 'Bearer ' . $token
        ];
    }

    /**
     * Search domain
     * @return Object
     */
    public function searchDomain($domain){
        $domain = [
            "domain" => $domain
        ];

        // Create search record
        $searchRecord = $this->_post('domain-searches', $domain);

        return $searchRecord;
    }

     /**
     * Create new Contact
     */
    public function createContact($contact = []){
        return $this->_post('contacts', $contact);
    }

    /**
     * Create admin contact
     */
     /**
     * Create new Contact
     */
    public function createAdminContact($contact = []){
        return $this->_post('domainadmins', $contact);
    }

    /**
     * Create new NS
     */
    public function createNS($ns){
        return $this->_post('nameservers', $ns);
    }

    /**
     * Create Domain
     */
    public function createDomain($data){
        return $this->_post('domains', $data);
    }

    /**
     * Register Domain
     */
    public function registerDomain($data){
        // Check if domain does not exist first
        $domain = null;
        $domain_name = $data['domain'];
        $domainFilter = [
            'name' => $domain_name
        ];
        
        $remoteDomain = $this->_get('domains', '', $domainFilter)->data;

        if (!empty($remoteDomain)) {
            $domain = $remoteDomain[0];
        }
        
        if (is_null($domain)){
             // Create Domain
            $domainData = [
                "name" => $data['domain'],
                "status" => "pending"
            ];
            $domain = $this->createDomain($domainData)->data;
    
            // Create Contact -registrant
            $contact_data = array_merge($data['contacts']['registrant'], ['domain_id' => $domain->id]);
            $contact = $this->createContact($contact_data);

            // Create Contact - admin
            $contact_data = array_merge($data['contacts']['admin'], ['domain_id' => $domain->id]);
            $contact = $this->createAdminContact($contact_data);
    
            // Create NS
            $ns_data = array_merge($data['nameservers'], ['domain_id' => $domain->id]);
            $ns = $this->createNS($ns_data);
        }
        
        $domain_id = $domain->id;
        
        $domainStatus = $domain->attributes->status;
        
        if ($domainStatus !== 'registered'){
            // Submit the domain for registration
            $registrationLink = 'domains/' . $domain_id . '/register';
            return $this->_post($registrationLink, $data);
        } else {
            throw new \Exception('ERR66 : Domain already registered');
        }
    }

    /**
     * Transfer Domain
     */
    public function transferDomain($data){
        // Check if domain does not exist first
        $domain = null;
        $domain_name = $data['domain'];
        $domainFilter = [
            'name' => $domain_name
        ];
        
        $remoteDomain = $this->_get('domains', '', $domainFilter)->data;

        if (!empty($remoteDomain)) {
            $domain = $remoteDomain[0];
        }
        
        if ( is_null($domain) ){
             // Create Domain
            $domainData = [
                "name" => $data['domain'],
                "status" => "pending"
            ];
            $domain = $this->createDomain($domainData)->data;
    
            // Create Contact
            $contact_data = array_merge($data['contacts']['registrant'], ['domain_id' => $domain->id]);
            $contact = $this->createContact($contact_data);

            // Create admin contact
            $contact_data = array_merge($data['contacts']['admin'], ['domain_id' => $domain->id]);
            $contact = $this->createAdminContact($contact_data);
    
            // Create NS
            $ns_data = array_merge($data['nameservers'], ['domain_id' => $domain->id]);
            $ns = $this->createNS($ns_data);
        }
        
        $domain_id = $domain->id;
        
        $domainStatus = $domain->attributes->status;
        
        if ($domainStatus !== 'transferred'){
            // Submit the domain for registration
            $transferLink = 'domains/' . $domain_id . '/transfer';
            return $this->_post($transferLink, $data);
        } else {
            throw new \Exception('ERR67 : Domain already transferred');
        }
    }
    
    public function getDomainInfo($data){
        $domain_name = $data['domain']['name'];
        $domainFilter = [
            'name' => $domain_name
        ];
        
        $domain = $this->_get('domains', '', $domainFilter)->data[0]->attributes;
        
        if( is_null ( $domain )){
            // Create domain if it does not exist : TOREMOVE
            $domainData = [
                'domain' => $domain_name,
                'status' => 'registered',
                'registration_date' => $data['domain']['registration_date'],
                'expiration_date' => $data['domain']['expiration_date']
            ];
            
            
            $domain = $this->createDomain($domainData);
            $domain_id = $domain->data->id;
            
            // Create Contact
            $contact_data = array_merge($data['contacts']['registrant'], ['domain_id' => $domain_id]);
            $contact = $this->createContact($contact_data);
    
            // Create NS
            $ns_data = array_merge($data['nameservers'], ['domain_id' => $domain_id]);
            $ns = $this->createNS($ns_data);
            
            return $domain->data->attributes;
        } else {
            return $domain;
        }
    }

    public function getWhoisInformation($postData){
        $domainFilter = [
           'name' => $postData['domain']
       ];
       
       $domain_id = $this->getDomainId($domainFilter);
       
       $contactsFilters = [
           "domain_id" => $domain_id
       ];
       
       $contacts =  $this->_get('contacts', "", $contactsFilters)->data[0]->attributes;
      
       return $contacts;
   }
    
    public function getNameservers($postData){
         $domainFilter = [
            'name' => $postData['domain']
        ];
        
        $domain_id = $this->getDomainId($domainFilter);
        
        $nameserversFilters = [
            "domain_id" => $domain_id
        ];
        
        $ns =  $this->_get('nameservers', "", $nameserversFilters)->data[0]->attributes;
       
        return $ns;
    }

    /**
     * Get Domains
     *
     */
     public function getDomains($filters){
        $domains = [];
        $response =  $this->_get('domains', "", $filters);

        if ( ! empty( $domains )) {
            foreach ($domains as $key => $domain) {
                array_push($domains, $domain->attributes);
            }
        }
       
        return $domains;
    }
    
    protected function getDomainId($domain){
        return $this->_get("domains", "", $domain)->data[0]->id;
    }
    
    public function setNameservers($postData){
        $domainFilter = [
            'name' => $postData['domain']
        ];
        
        $domain_id = $this->getDomainId($domainFilter);
        
        $nameserversFilters = [
            "domain_id" => $domain_id
        ];
        
        $nsId = $this->_get("nameservers", "", $nameserversFilters)->data[0]->id;
        
        return $this->_patch('nameservers', $nsId, $postData['nameservers']);
    }

    public function updateWhoisInformation($data){
        $domainFilter = [
            'name' => $data['domain']
        ];
        
        $domain_id = $this->getDomainId($domainFilter);
        

        // Get Contact
        $contactsFilter = [
            "domain_id" => $domain_id
        ];

        $contantId = $this->_get('contacts', '', $contactsFilter)->data[0]->id;
        $registrant = $this->_patch('contacts', $contantId, $data['contacts']['registrant']);
        
        // // Get registrant
        // $registrantFilter = [
        //     "contact_id" => $contantId
        // ];

        // $registrantId = $this->_get('registrants', '', $registrantFilter)->data[0]->id;

        // // Update registrant
        // $registrant = $this->_patch('registrants', $registrantId, $data['contacts']['registrant']);

        // // Get tech contact
        // $techFilter = [
        //     "contact_id" => $contantId
        // ];

        // $techId = $this->_get('techs', '', $techFilter)->data[0]->id;

        // // Update Tech Contact
        // $tech = $this->_patch('techs', $techId, $data['contacts']['tech']);

        return [
            'contacts' => [
                'registrant' => $registrant,
                // 'tech' => $tech
            ]
        ];
    }

    /**
     * Login to API
     */
    public function authorize(){
        // $password = getenv('PASSWORD');
        // $email = getenv('EMAIL');
        $client_id = getenv('CLIENT_ID');
        $secret = getenv('SECRET');

        $options = [
            'form_params' =>  [
                "grant_type" => "client_credentials",
                "client_id" => $client_id,
                "client_secret" => $secret
            ]
        ];

        try
        {
            $response = $this->client->post('/oauth/token', $options);

            $responseString = (string) $response->getBody();

            $responseArray = json_decode($responseString);

            $this->token = $responseArray->access_token;
        } 
        catch(\Exception $ex){
            // echo $ex->getMessage();
            $error = [
                "errors" => [
                    [
                        "status" => "401",
                        "title" => "Failed to login"
                    ]
                ]
            ];
            
            return json_encode($error);
        }
    }

    protected function _get($resource, $record = "", $filters = []){
        $options = [
            "headers" => $this->headers,
            "http_errors" => false,
            'timeout' => 2000
        ];
        
        $url = $this->apiUrl . '/' . $resource;
        
        if( ! empty($record)){
            $url .= "/" . $record;
        } elseif ( ! empty ($filters) ) {
            $url .= "?";
            
            foreach ( $filters as $key => $filter ) {
                $url .= "filter[$key]=$filter&";
            }
        }
        
        $response = $this->client->get($url, $options);
        
        $code = $response->getStatusCode();
        
        if ($code == '200'){
            return json_decode( ( string ) $response->getBody());
        } else {
            //echo (( string ) $response->getBody());
            
            // die($url);
            throw new \Exception('ERR' . $code . ' : Error fetching data for domain');
        }
    }

    protected function _post($resource, $data){
        $url = $this->apiUrl . '/' . $resource;

        $options = [
            "json" => [
                "data" => [
                    "type" => $resource,
                    "attributes" => $data
                ]
            ],
            "headers" => $this->headers,
            "http_errors" => false
        ];
        
         $response = $this->client->post(
            $url, 
            $options
        );
        
        $code = $response->getStatusCode();
            
        if ($code == '401'){
            throw new \Exception('Authentiation to registrar failed. Verify your token');
        } elseif ($code == '500'){
            throw new \Exception('ERR43 : Domain could not be registered');
        } elseif ($code == '201' || $code == '200'){
            // Successful but with custom errors
            $data = json_decode((string) $response->getBody());
            
//            if (! is_null($data->errors)){
  //              $errors = implode(',', $data->errors);
    //            
      //          throw new \Exception($errors);
        //    }
            
            return $data;
        }

    }

    protected function _patch($resource, $record, $data){
        $url = $this->apiUrl . '/' . $resource . '/' . $record;

        $options = [
            "json" => [
                "data" => [
                    "type" => $resource,
                    "id" => "$record",
                    "attributes" => $data
                ]
            ],
            "headers" => $this->headers,
             "http_errors" => false
        ];
        
        $response = $this->client->patch(
            $url, 
            $options
        );
        
        $code = $response->getStatusCode();
        
        
        
        if ($code == '200' || $code == '201'){
            return json_decode( ( string ) $response->getBody());
        } else {
            // print_r($data);
            // echo (( string ) $response->getBody());
            
            // die($url);
            throw new \Exception('ERR' . $code . ' : Error updating data');
        }
    }
}

